Makefile

```diff
@@ -11,65 +11,109 @@ else
+# file as xyz: ## something, and then pretty-format the target and help. Then,
+# if there's a line with ##@ something, that gets pretty-printed as a category.
+# More info on the usage of ANSI control characters for terminal formatting:
+# https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters
+# More info on the awk command:
+# http://linuxcommand.org/lc3_adv_awk.php
 
-# Build manager binary
-manager: generate fmt vet
-	go build -o bin/manager main.go
+help: ## Display this help.
+	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_0-9-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
 
-# Run against the configured Kubernetes cluster in ~/.kube/config
-run: generate fmt vet manifests
-	go run ./main.go --zap-devel=true
+##@ Development
 
-debug: fmt vet
-	./hack/kill-debug.sh
-	./hack/debug.sh
+manifests: controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
+	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
 
-# Install CRDs into a cluster
-install: manifests
-	kustomize build config/crd | kubectl apply -f -
+generate: controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
+	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
 
-# Uninstall CRDs from a cluster
-uninstall: manifests
-	kustomize build config/crd | kubectl delete -f -
+fmt: ## Run go fmt against code.
+	go fmt ./...
 
-# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
-deploy: manifests
-	cd config/manager && kustomize edit set image controller=${IMG}
-	kustomize build config/default | kubectl apply -f -
+shftm: shfmt ## Run shfmt against code.
+	$(SHFMT) -l -d hack/.
 
-# Generate manifests e.g. CRD, RBAC etc.
-manifests:
-	controller-gen $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+vet: ## Run go vet against code.
+	go vet ./...
 
-# Run go fmt against code
-fmt:
-	go fmt ./...
+ENVTEST_ASSETS_DIR=$(shell pwd)/testbin
+test: manifests generate fmt shfmt vet ginkgo ## Run tests.
+	mkdir -p ${ENVTEST_ASSETS_DIR}
+	test -f ${ENVTEST_ASSETS_DIR}/setup-envtest.sh || curl -sSLo ${ENVTEST_ASSETS_DIR}/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.8.3/hack/setup-envtest.sh
+	source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR); go test ./... -coverprofile cover.out
 
-# Run go vet against code
-vet:
-	go vet ./...
+##@ Build
 
-# Generate code
-generate:
-	controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
+build: generate fmt vet ## Build manager binary.
```

Comment:
I think it would make sense to add `test` in the list?

Reply:
This is coming from kubebuilder but I can add that

Reply:
So those are our two targets:
- `test: manifests generate fmt shfmt vet ginkgo ## Run tests.`
- `build: generate fmt vet test ## Build manager binary.`
It seems to me that if we add test to build we have a lot of repetitions. What do you think?